spring-kafka/src/main/java/org/springframework/kafka/listener/ConcurrentMessageListenerContainer.java

```diff
@@ -374,24 +376,38 @@ protected void doStop(final Runnable callback, boolean normal) {
 		}
 	}
 
+	@Override
+	public void childStarted(MessageListenerContainer child) {
+		this.startedContainers.incrementAndGet();
+	}
+
 	@Override
 	public void childStopped(MessageListenerContainer child, Reason reason) {
 		if (this.reason == null || reason.equals(Reason.AUTH)) {
 			this.reason = reason;
 		}
-		if (Reason.AUTH.equals(this.reason)
-				&& getContainerProperties().isRestartAfterAuthExceptions()
-				&& this.concurrency == this.stoppedContainers.incrementAndGet()) {
+		int startedContainersCount = this.startedContainers.decrementAndGet();
+		if (startedContainersCount == 0) {
+			publishConcurrentContainerStoppedEvent();
```

Comment:
I wonder if we really should emit this event here unconditionally.
Looks like in case of `Reason.AUTH` we restart the container immediately.
Therefore we go to the state when children are running, so `ConcurrentContainerStoppedEvent` is loosing its purpose.

Reply:
**ConcurrentContainer restart happens only when all the containers are stopped**. Whenever all the containers are stopped, then as per the contract ConcurrentContainerStopped event has to be published. 

Please give your comments.

Reply:
That's true, but are they really stopped if we restart it immediately?
That's my concern if we should rethink this situation and don't emit that `ConcurrentContainerStoppedEvent` if we restart immediately.

According to your original request about stopping the whole application context on that event, such a restart might cause some problem.
So, why don't we change the logic to be sure that this is exactly a moment for `ConcurrentContainerStoppedEvent`?

Reply:
I got your point. I am good to make change so that `ConcurrentContainerStoppedEvent` is not published when AUTH error occurred. But, I am having one concern and one suggestion.

Concern:
	Whatever may be the scenario, as per the contract it is good to publish `ConcurrentContainerStoppedEvent` when all the containers are stopped.
	
Suggestion:
	Error reason could be added to `ConcurrentContainerStoppedEvent`, so that application user would take a right decision.

public ConcurrentContainerStoppedEvent(Object source, Reason reason)


Please give your idea.

Reply:
> but are they really stopped if we restart it immediately?

Yes, childStopped API will be called only when the container  stopped.

Reply:
Yes, that sounds good to have that `Reason` propagated down to the `ConcurrentContainerStoppedEvent`.
Thank you!

Reply:
shall we propagate all reasons for each container as below. As of now, only Auth reason is propagated. Is that sufficient.

public ConcurrentContainerStoppedEvent(Object source, ConsumerStoppedEvent.Reason[] reason) -- all reasons will be propagated.

or 

public ConcurrentContainerStoppedEvent(Object source, Reason reason) -- only Auth reason propagated. Other will not be propagated.

Reply:
If we propagate only `Reason.AUTH`, then no reason in this `reason` property for the event.
I don't think we have to propagate all of them.
The logic is like this:
```
	if (this.reason == null || reason.equals(Reason.AUTH)) {
			this.reason = reason;
		}
```
So, any first reason is going to be use (or AUTH) when it overrides the value of that property.